src/mongo_odm/model.hpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/oid.hpp>
+#include <mongo_odm/config/prelude.hpp>
+#include <mongo_odm/odm_collection.hpp>
+#include <mongocxx/collection.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+template <typename T, typename IdType = bsoncxx::oid>
+class model {
+   private:
+// TODO: When XCode 8 is released, this can always be thread_local. Until then, the model class
+//       will not be thread-safe on OS X.
+#ifdef __APPLE__
+    static odm_collection<T> _coll;
+#else
+    static thread_local odm_collection<T> _coll;
+#endif
+
+   protected:
+    // TODO: if IdType is bsoncxx::oid, the derived class must initialize this until CXX-940 is
+    //       resolved.
+    //
+    // If IdType is not bsoncxx::oid, the derived class is always responsible for initializing it.
+    IdType _id;
+
+   public:
+    /**
+     * Returns a copy of the underlying collection.
+     *
+     * @return A copy of the underlying mongocxx::collection that this class uses to store and
+     * load instances of T.
+     */
+    static const mongocxx::collection& collection() {
```

Comment:
On closer examination and more thought here, I think returning by value (and leaving the caller with a copy) is actually the best thing to do (sorry to make you switch it again). The underlying odm_collection is returning by value, so we could be making a copy here anyway. Collection objects are pretty small, so I think it's best not to worry about it.